src/event/pmix_event_notification.c

```diff
@@ -190,124 +192,189 @@ static void progress_local_event_hdlr(pmix_status_t status,
+                if (nxt->codes[n] == chain->status) {
+                    chain->evhdlr = nxt;
                     /* add any cbobject - the info struct for it is at the end */
-                    chain->info[chain->ninfo-1].value.data.ptr = multi->cbobject;
-                    multi->evhdlr(multi->index,
-                                  chain->status, &chain->source,
-                                  chain->info, chain->ninfo,
-                                  chain->results, chain->nresults,
-                                  progress_local_event_hdlr, (void*)chain);
-                    goto complete;
+                    chain->info[chain->ninfo-1].value.data.ptr = nxt->cbobject;
+                    nxt->evhdlr(nxt->index,
+                                chain->status, &chain->source,
+                                chain->info, chain->ninfo,
+                                chain->results, chain->nresults,
+                                progress_local_event_hdlr, (void*)chain);
+                    return;
                 }
             }
         }
         /* if we get here, then there are no more multi-mode
          * events that match */
-        chain->multi = NULL;
-        /* pickup the beginning of the default event list */
-        chain->def = (pmix_default_event_t*)pmix_list_get_begin(&pmix_globals.events.default_events);
+        item = pmix_list_get_begin(&pmix_globals.events.default_events);
     }
 
-    /* if they didn't want it to go to a default handler, then we are done */
-    if (chain->nondefault) {
-        goto complete;
+    /* if they didn't want it to go to a default handler, then ignore them */
+    if (!chain->nondefault) {
+        if (NULL == item) {
+            item = &chain->evhdlr->super;
+        }
+        if (pmix_list_get_end(&pmix_globals.events.default_events) != (item = pmix_list_get_next(item))) {
+            nxt = (pmix_event_hdlr_t*)item;
+            /* if this event handler provided a range, check to see if
+             * the source fits within it */
+            if (check_range(&nxt->rng, &chain->source)) {
+                chain->evhdlr = nxt;
+                /* add any cbobject - the info struct for it is at the end */
+                chain->info[chain->ninfo-1].value.data.ptr = nxt->cbobject;
+                nxt->evhdlr(nxt->index,
+                            chain->status, &chain->source,
+                            chain->info, chain->ninfo,
+                            chain->results, chain->nresults,
+                            progress_local_event_hdlr, (void*)chain);
+                return;
+            }
+        }
     }
 
-    if (NULL != chain->def) {
-        if (pmix_list_get_end(&pmix_globals.events.default_events) != (nxt = pmix_list_get_next(&chain->def->super))) {
-            def = (pmix_default_event_t*)nxt;
-            chain->def = def;
+    /* if we registered a "last" handler, and it fits the given range
+     * and code, then invoke it now */
```

Comment:
Yes - remember, there can be any number of registered "default" handlers. All "default" means is that the handler shall be invoked for any code.